### PR TITLE
Generate e2e benchmark model tests from python framework

### LINKED ISSUE #10988: Migrate e2e benchmark module tests to e2e test framework
The follow-up of #10442 to migrate e2e module tests from the current benchmark suites to e2e test framework.

### PR DESCRIPTION
**Pull request by pzread**
Defines and generates e2e benchmark models with the e2e test python framework.

Fixes #10988

### CODE REVIEW

File: tests/e2e/models/generated_e2e_model_tests.cmake [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  EXPECTED_OUTPUT
+    "efficientnet_int8_expected_output.txt"
+  RUNNER_ARGS
+    "--entry_function=main"
+    "--function_input=1x224x224x3xui8=0"
+)
+
+iree_benchmark_suite_module_test(
+  NAME
+    "deeplab_v3_fp32_correctness_test"
```
Comment (hcindyl) [inline tests/e2e/models/generated_e2e_model_tests.cmake:53]: https://github.com/iree-org/iree/issues/10748. This may not work.

File: tests/e2e/models/generated_e2e_model_tests.cmake [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  MODEL
+    "78eab9e5-9ff1-4769-9b55-933c81cc9a0f_MobileNetV1_fp32"
+  DRIVER
+    "local-sync"
+  EXPECTED_OUTPUT
+    "mobilenet_v1_fp32_expected_output.txt"
+  RUNNER_ARGS
+    "--entry_function=main"
+    "--function_input=1x224x224x3xf32=0"
+)
```
Comment (hcindyl) [inline tests/e2e/models/generated_e2e_model_tests.cmake:35]: Still need to mark `UNSUPPORTED_PLATFORMS`. For this target, it is riscv-rv32-Linux

File: tests/e2e/models/generated_e2e_model_tests.cmake (hunk 1 above)
Comment (pzread) [reply tests/e2e/models/generated_e2e_model_tests.cmake:58]: Looks like #10748 has been resolved (https://github.com/iree-org/iree/commit/cc183199c11df2ce3cd8696251ca2c30210b36e9)

File: tests/e2e/models/generated_e2e_model_tests.cmake (hunk 2 above)
Comment (pzread) [reply tests/e2e/models/generated_e2e_model_tests.cmake:38]: Done.

File: tests/e2e/models/generated_e2e_model_tests.cmake [hunk 3]
```diff
@@ -0,0 +1,88 @@
+################################################################################
```
Comment (hcindyl) [inline tests/e2e/models/generated_e2e_model_tests.cmake:1]: Similar to #10928, do we want to check-in the generated .cmake file? The .cmake file can be called with
```
if(DEFINED IREE_BENCHMARK_SUITE_DIR)
  file(GENERATE ...)
  include(${CMAKE_CURRENT_BINARY_DIR}/generated_e2e_model_tests.cmake)
endif()
```

File: build_tools/cmake/iree_run_module_test.cmake [hunk 4]
```diff
@@ -215,9 +213,8 @@ endfunction()
 #
 # Parameters:
 #   NAME: Name of the target
-#   BENCHMARK_MODULE_SRC: IREE module flagfile path built from benchmark_suite.
-#       The flagfile for different compile configurations are stored in the
-#       subdirectories.
+#   MODEL: Model "<UUID>_<name>" defined in
+#       "build_tools/python/e2e_test_framework/models".
```
Comment (hcindyl) [inline build_tools/cmake/iree_run_module_test.cmake:217]: I can't find UUID information in https://github.com/iree-org/iree/blob/main/build_tools/python/e2e_test_framework/models. Could we be more specific?

File: tests/e2e/models/generated_e2e_model_tests.cmake (hunk 3 above)
Comment (pzread) [reply tests/e2e/models/generated_e2e_model_tests.cmake:1]: It's actually not that simple because `file(GENERATE...` doesn't run at configuration step, so the `include(...` won't be able to find the file.

There are some more involved ways to do that (https://stackoverflow.com/questions/52556785/code-generator-generating-its-own-cmake-files-and-targets) But I haven't give them a try.

Tag #10928 to make sure we won't forget this one.

File: build_tools/cmake/iree_run_module_test.cmake (hunk 4 above)
Comment (pzread) [reply build_tools/cmake/iree_run_module_test.cmake]: Added some more description.

File: build_tools/cmake/build_benchmarks.sh [hunk 5]
```diff
@@ -33,8 +33,8 @@ echo "Configuring to build benchmarks"
 "${CMAKE_BIN}" -B "${BUILD_BENCHMARKS_DIR}" \
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_benchmarks.sh:33]: Here for threading:

Did you meant to include "Fixes" pointing at a pull request?

File: build_tools/cmake/build_benchmarks.sh [hunk 6]
```diff
@@ -33,8 +33,8 @@ echo "Configuring to build benchmarks"
...
   -G Ninja \
   -DIREE_HOST_BINARY_ROOT="${IREE_HOST_BINARY_ROOT}" \
-  -DIREE_BUILD_BENCHMARKS=ON \
   -DIREE_BUILD_MICROBENCHMARKS=ON \
+  -DIREE_BUILD_EXPERIMENTAL_PYTHON_GENERATED_BENCHMARKS=ON \
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_benchmarks.sh:37]: This is a big step! Let's make sure we're ready before flipping this. Or perhaps this is just for testing during presubmit?

File: build_tools/cmake/iree_run_module_test.cmake [hunk 7]
```diff
@@ -7,9 +7,7 @@
 # Utility function to return the platform name in crosscompile.
 function(iree_get_platform PLATFORM)
   if(ANDROID AND CMAKE_ANDROID_ARCH_ABI STREQUAL "arm64-v8a")
-    set(_PLATFORM "android-arm64-v8a")
-  elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-    set(_PLATFORM "x86_64")
+    set(_PLATFORM "arm64-v8a-Android")
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_run_module_test.cmake:10]: How is this related?

File: build_tools/cmake/iree_run_module_test.cmake [hunk 8]
```diff
@@ -276,48 +274,20 @@ function(iree_benchmark_suite_module_test)
...
-  # TODO(#10391): Update this logic with the new benchmark framework.
-  foreach(_FLAG ${_FLAGS})
-    if(_FLAG MATCHES "--module_file=")
-      string(REPLACE "--module_file=" "" _SRC "${_FLAG}")
-      set(_SRC "${_FLAG_FILE_DIR}/${_SRC}")
-    endif()
-  endforeach(_FLAG)
-
+  # Drop the "<UUID>_" prefix.
+  string(SUBSTRING "${_RULE_MODEL}" 37 -1 _MODEL_NAME)
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_run_module_test.cmake:285]: I think a bit more of a comment would be helpful (pointing to the UUID standard and noting that they're always 36 characters, for instance)

File: build_tools/cmake/build_benchmarks.sh (hunk 6 above)
Comment (pzread) [reply build_tools/cmake/build_benchmarks.sh:37]: `build_tools/cmake/build_benchmarks.sh` currently is only used in CI pipeline `build_suites` and for e2e model tests. So I think it's safe to flip now with the changes of e2e model tests. The downside is the old benchmark suites and the new one might be out of sync and confusing (the buildkite pipeline and CI are using different benchmark suites).

I'm fine to flip this later there is a concern.

The buildkite benchmark pipeline is using a separate script to build benchmark modules.

Comment (pzread) [reply build_tools/cmake/build_benchmarks.sh:37]: But if we don't turn this on, it'll be hard to test the new benchmark pipelines, unless we are okay to build both old and new benchmark suites in CI at the same time.

File: build_tools/cmake/iree_run_module_test.cmake (hunk 7 above)
Comment (pzread) [reply build_tools/cmake/iree_run_module_test.cmake:10]: The original ad-hoc naming is to fit the names in old benchmark suites. I canonicalized them since we don't need to match old benchmark suites anymore.

File: build_tools/cmake/build_benchmarks.sh (hunk 5 above)
Comment (pzread) [reply build_tools/cmake/build_benchmarks.sh:33]: Realized it's not an issue. Filed and pointed to a new one #10988

File: build_tools/cmake/iree_run_module_test.cmake (hunk 8 above)
Comment (pzread) [reply build_tools/cmake/iree_run_module_test.cmake:285]: Done.

File: build_tools/cmake/iree_run_module_test.cmake (hunk 7 above)
Comment (GMNGeoffrey) [reply build_tools/cmake/iree_run_module_test.cmake:10]: I guess I'm just confused why android is still special but also gets a new name. Does it not set `CMAKE_SYSTEM_PROCESSOR` and `CMAKE_SYSTEM_NAME`? Maybe we should do that

Comment (hcindyl) [reply build_tools/cmake/iree_run_module_test.cmake:10]: android.toolchain.cmake use its own variables, although they are supported by CMake.
https://android.googlesource.com/platform/ndk/+/master/build/cmake/android.toolchain.cmake#116

Comment (pzread) [reply build_tools/cmake/iree_run_module_test.cmake:10]: It looks like CMake has an internal module to guess the `CMAKE_SYSTEM_PROCESSOR` for Android
https://github.com/Kitware/CMake/blob/master/Modules/Platform/Android-Determine.cmake
But for arm64-v8a, CMake will set `CMAKE_SYSTEM_PROCESSOR` as `aarch64`.

Comment (pzread) [reply build_tools/cmake/iree_run_module_test.cmake]: Simplified to just use `${CMAKE_SYSTEM_PROCESSOR}-${CMAKE_SYSTEM_NAME}`. We can revise this if there is an issue to use `aarch64` in CMake.

File: tests/e2e/models/CMakeLists.txt [hunk 9]
```diff
@@ -147,71 +147,4 @@ iree_static_linker_test(
...
-    "local-sync"
-  RUNNER_ARGS
-    "--entry_function=main"
-    "--function_input=1x96x96x1xi8=0"
-  EXPECTED_OUTPUT
-    "1x2xi8=[72 -72]"
-  UNSUPPORTED_PLATFORMS
-    "android-arm64-v8a"
-)
+include(generated_e2e_model_tests.cmake)
```
Comment (hcindyl) [inline tests/e2e/models/CMakeLists.txt:150]:
```suggestion
if(DEFINED IREE_BENCHMARK_SUITE_DIR)
  # TODO(#10928): Add file generation step. 
  include(generated_e2e_model_tests.cmake)
endif()
```

File: build_tools/cmake/build_benchmarks.sh (hunk 6 above)
Comment (hcindyl) [reply build_tools/cmake/build_benchmarks.sh:37]: We use the buildkite benchmark result for performance regression, so having GH presubmit and buildkite out of sync can be concerning.

Maybe we can run this path in postsubmit only, create a temporary fork `iree_experimental_benchmark_suite_module_test` with the CMake implementation change, and keep the current `tests/e2e/model/CMakeLists.txt` target until we see everything is stable and we are ready to flip both github action and buildkite?

**Review by hcindyl — COMMENTED** [review]
I am okay with the code change. However, I understand @GMNGeoffrey 's concern of flipping the flag for good.

File: build_tools/cmake/build_benchmarks.sh (hunk 6 above)
Comment (pzread) [reply build_tools/cmake/build_benchmarks.sh:37]: Creating a fork in postsubmit pipeline to build the new benchmark suite will still cause the benchmark suite built twice in postsubmit, which increases the loading of runners. I'm okay to do so if we can accept the overhead.

Although personally I don't see a very high risk to flip this right now except for the concern of being out of sync. Because in presubmit we can test if the e2e model tests and buildkite pipeline works correctly.

Comment (GMNGeoffrey) [reply build_tools/cmake/build_benchmarks.sh:37]: I'd rather not have the skew. That means we could discover a problem only on postsubmit. Let's start with the duplication, migrate to the new system, and then turn down the old one

File: build_tools/python/e2e_model_tests/test_definitions.py [hunk 10]
```diff
@@ -18,7 +18,7 @@
 
 class CMakePlatform(enum.Enum):
   """Enum of CMake system platform string."""
-  ANDROID_ARMV8_A = "arm64-v8a-Android"
+  ANDROID_ARMV8_A = "aarch64-Android"
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/test_definitions.py:21]: I still don't understand what the difference between arm64 and aarch64 is. I think the v8a is maybe relevant though?

Comment (GMNGeoffrey) [reply build_tools/python/e2e_model_tests/test_definitions.py:21]: Probably @antiagainst would know best

**Review by GMNGeoffrey — APPROVED** [review]

File: build_tools/cmake/build_benchmarks.sh (hunk 6 above)
Comment (pzread) [reply build_tools/cmake/build_benchmarks.sh]: Sure, I'm happy that we agreed to build the benchmark suite twice during the migration : ) 
Removed any CI and test changes from this PR. Only adds the tool to generate the CMake files.
